src/subclass/input_stream.rs

```diff
@@ -0,0 +1,322 @@
+                .read_fn
+                .expect("No parent class implementation for \"read\"");
+            let mut err = ptr::null_mut();
+            let res = f(
+                stream.to_glib_none().0,
+                buffer.as_mut_ptr() as glib_sys::gpointer,
+                buffer.len(),
+                cancellable.to_glib_none().0,
+                &mut err,
+            );
+            if res == -1 {
+                Err(from_glib_full(err))
+            } else {
+                Ok(res as usize)
+            }
+        }
+    }
+
+    fn parent_close(
+        &self,
+        stream: &InputStream,
+        cancellable: Option<&Cancellable>,
+    ) -> Result<(), Error> {
+        unsafe {
+            let data = self.get_type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut gio_sys::GInputStreamClass;
+            let mut err = ptr::null_mut();
+            if let Some(f) = (*parent_class).close_fn {
+                if from_glib(f(
+                    stream.to_glib_none().0,
+                    cancellable.to_glib_none().0,
+                    &mut err,
+                )) {
+                    Ok(())
+                } else {
+                    Err(from_glib_full(err))
+                }
+            } else {
+                Ok(())
+            }
+        }
+    }
+
+    fn parent_skip(
+        &self,
+        stream: &InputStream,
+        count: usize,
+        cancellable: Option<&Cancellable>,
+    ) -> Result<usize, Error> {
+        unsafe {
+            let data = self.get_type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut gio_sys::GInputStreamClass;
+            let mut err = ptr::null_mut();
+            let f = (*parent_class)
+                .skip
+                .expect("No parent class implementation for \"skip\"");
+            let res = f(
+                stream.to_glib_none().0,
+                count,
+                cancellable.to_glib_none().0,
+                &mut err,
+            );
+            if res == -1 {
+                Err(from_glib_full(err))
+            } else {
+                Ok(res as usize)
+            }
+        }
+    }
+}
+
+unsafe impl<T: ObjectSubclass + InputStreamImpl> IsSubclassable<T> for InputStreamClass {
+    fn override_vfuncs(&mut self) {
+        <glib::ObjectClass as IsSubclassable<T>>::override_vfuncs(self);
+        unsafe {
+            let klass = &mut *(self as *mut Self as *mut gio_sys::GInputStreamClass);
+            klass.read_fn = Some(stream_read::<T>);
+            klass.close_fn = Some(stream_close::<T>);
+            klass.skip = Some(stream_skip::<T>);
+        }
+    }
+}
+
+unsafe extern "C" fn stream_read<T: ObjectSubclass>(
+    ptr: *mut gio_sys::GInputStream,
+    buffer: glib_sys::gpointer,
+    count: usize,
+    cancellable: *mut gio_sys::GCancellable,
+    err: *mut *mut glib_sys::GError,
+) -> isize
+where
+    T: InputStreamImpl,
+{
+    use std::slice;
```

Comment:
Why did you put this import here?

Reply:
Because it's only used in this single function. Didn't want to pollute the global imports

Reply:
Well, not sure imports are supposed to work this way but I don't really mind...